common/lib/xmodule/xmodule/modulestore/split_mongo/mongo_connection.py

```diff
@@ -1,50 +1,153 @@
-import pytz
+import dogstats_wrapper as dog_stats_api
 
 
 new_contract('BlockData', BlockData)
 
 
-def structure_from_mongo(structure):
+def round_power_2(value):
+    """
+    Return value rounded up to the nearest power of 2.
+    """
+    if value == 0:
+        return 0
+
+    return math.pow(2, math.ceil(math.log(value, 2)))
+
+
+class Tagger(object):
+    def __init__(self):
+        self.added_tags = []
+        self.measures = []
+
+    def measure(self, name, size):
+        self.measures.append((name, size))
+
+    def tag(self, **kwargs):
+        self.added_tags.extend(kwargs.items())
+
+    @property
+    def tags(self):
+        return [
+            '{}:{}'.format(name, round_power_2(size))
+            for name, size in self.measures
+        ] + [
+            '{}:{}'.format(name, value)
+            for name, value in self.added_tags
+        ]
+
+
+class QueryTimer(object):
+    def __init__(self, metric_base, sample_rate=1):
+        """
+        Arguments:
+            metric_base: The prefix to be used for all queries captured
+            with this :class:`QueryTimer`.
+        """
+        self._metric_base = metric_base
+        self._sample_rate = sample_rate
+
+    @contextmanager
+    def timer(metric_name, course_context):
+        """
+        Contextmanager which acts as a timer for the metric ``metric_name``,
+        but which also yields a :class:`Tagger` object that allows the timed block
+        of code to add tags and quantity measurements. Tags are added verbatim to the
+        timer output. Measurements are recorded as histogram measurements in their own,
+        and also as bucketed tags on the timer measurement.
+
+        Arguments:
+            metric_name: The name used to aggregate all of these metrics.
+            course_context: The course which the query is being made for.
+        """
+        tagger = Tagger()
+        metric_name = "{}.{}".format(self._metric_base, metric_name)
+
+        start = time()
+        try:
+            yield tagger
+        finally:
+            end = time()
+            tags = tagger.tags
+            tags.append('course:{}'.format(course_context))
+            for name, size in tagger.measures:
+                dog_stats_api.histogram(
+                    '{}.{}'.format(metric_name, name),
+                    size,
+                    timestamp=end,
+                    tags=[tag for tag in tags if not tag.startswith('{}:'.format(metric_name))],
+                    sample_rate=self._sample_rate,
+                )
+            dog_stats_api.histogram(
+                '{}.duration'.format(metric_name),
+                end - start,
+                timestamp=end,
+                tags=tags,
+                sample_rate=self._sample_rate,
+            )
+            dog_stats_api.increment(
+                metric_name,
+                timestamp=end,
+                tags=tags,
+                sample_rate=self._sample_rate,
+            )
+
```

Comment:
But that doesn't change the number of calls to dog_stats_api, right? It just changes the amount of data that's queued up and ultimately sent to DataDog?